<h1><center>Laboratorio 2: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Pablo Badilla
- Auxiliar: Ignacio Meza D.
- Ayudante: Diego Irarrázaval

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Diego Canales


### **Link de repositorio de GitHub:** `https://github.com/DiegoCanalesR/MDS7202`

### Indice 

1. [Temas a tratar](#Temas-a-tratar:)
3. [Descripcción del laboratorio](#Descripción-del-laboratorio.)
4. [Desarrollo](#Desarrollo)

# Temas a tratar

- Manejo de DataFrames con la libreria `pandas`.
- Visualizaciones interactivas de los datos con `plotly`.

## Reglas:

- Fecha de entrega: TBD
- **Grupos de 2 personas**
- **Ausentes** deberán realizar la actividad solos. 
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquer material y segmentos de código del curso que estimen conveniente.

### Objetivos principales del laboratorio

- Aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas`.
- Utilizar `plotly` para obtener información gráfica del dataset.
- Aplicar el **Análisis Exploratorio de Datos** a un caso en particular.


El laboratorio deberá ser desarrollado **sin el uso indiscriminado de nativos de python** (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

# Importamos librerias utiles 😸

In [1]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para visualizar
#!pip install --upgrade plotly
import plotly.express as px

# 1. Rendimiento en Estudiantes 📚


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

Para este laboratorio deberán realizar un Análisis Exploratorio de datos sobre el conjunto ```students_grades.csv```, el cual contiene una caracterización sobre el rendimiento y otros atributos de cada alumno de la *Universidad de la Cachaña* .

In [ ]:
# Usar solamente si utilizan Collab.
from google.colab import drive
drive.mount("/content/drive")
dir = '//content//drive//My Drive//CuerpoDocente//Laboratorios//Lab2//'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [240]:
df_grades = pd.read_csv('students_grades.csv')


## 1.1 Revisión de datos [1 punto]

<p align="center">
  <img src="https://media2.giphy.com/media/4FQMuOKR6zQRO/giphy.gif?cid=ecf05e47ii9lhu9tc9tb31d6lafrysp2zyf6u8w3y348mk8t&rid=giphy.gif&ct=g">
</p>



El Rector *Don Caguayo* le comenta que actualmente la Universidad posee un total de **1000** alumnos inscritos. Sin embargo, al revisar su *Excel*, no le cabe en la cabeza que hayan más de **1000** alumnos en la universidad. 

Consternado, el rector cree que se han realizado malas prácticas al inscribir tanto los registros de las notas como las notas como tal y les solicita a ustedes (aka, los expertos) que revisen los datos. 

Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**.

Cree una función que le permita revisar si las diferentes variables que componen el dataset son coherentes y si es posible observar anomalías en estas.

> ¿Sera necesario limpiar los datos?, comente esto en la celda ubicada debajo de **justificación**. (agregar si están bien el rango de notas)

**To-Do**

Crear la función que `analisis_exploratorio` que:
- [X] Imprima un extracto del Dataframe (usar `display` en vez de `print` para los DataFrames).
- [X] Imprima las dimensiones del DataFrame.
- [X] Describa los datos numéricos a través de ```.describe()```.
- [X] Imprima el total de filas repetidas (hint: Buscar el método```duplicated()```.).
- [X] Imprima la cantidad de valores nulos (`ìsna()`).
- [X] Imprima la cantidad de valores unicos.

Además, genere visualizaciones (fuera de la función) que:
- [X] Muestre la distribución de las variables numericas.

Por último, 

- [X] Justifique si será necesario limpiar los datos y aplique los análisis univariado y/o bivariado vistos en clase.


In [58]:
def analisis_preliminar_datos(df_grades):
    #Fragmento aleatorio del DataFrame
    print('--Fragmento del DataFrame:')
    display(df_grades.sample(5))
    print()
    
    #Dimensiones del DataFrame
    print('--Dimensiones del DataFrame:')
    print(f'{df_grades.shape[0]} filas y {df_grades.shape[1]} columnas')
    print()
    
    #Descripción de datos numéricos
    print('--Descripción de datos numéricos:')
    print(df_grades.describe())
    print()
    
    #Filas repetidas
    print('--Total de filas repetidas:')
    print(df_grades.duplicated().sum())
    print()
    
    #Valores nulos
    print('--Cantidad de valores nulos:')
    print('Por columna:')
    print(df_grades.isna().sum())
    print(f'Total: {df_grades.isna().sum().sum()}')
    print()
    
    #Valores únicos
    print('--Cantidad de valores únicos por columna:')
    print(df_grades.nunique())
    print()
  

In [59]:
analisis_preliminar_datos(df_grades)

--Fragmento del DataFrame:


names  gender race/ethnicity parental level of education  \
323  Allison Tankersley  female        group C                some college   
433     Greg Westerberg    male        group D          associate's degree   
649         Gary Hunter    male        group D          associate's degree   
432       Henry Mcgowen    male        group E                some college   
214       Autumn Miller  female        group D                 high school   

            lunch test preparation course  math score  reading score  \
323  free/reduced                    none        62.0           72.0   
433      standard               completed        81.0           72.0   
649      standard               completed        87.0           84.0   
432      standard                    none        76.0           71.0   
214      standard                    none        51.0           66.0   

     writing score  
323           70.0  
433           77.0  
649           85.0  
432           72.0  
214           62.0


--Dimensiones del DataFrame:
1400 filas y 9 columnas

--Descripción de datos numéricos:
        math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000

--Total de filas repetidas:
3

--Cantidad de valores nulos:
Por columna:
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64
Total: 127

--Cantidad de valores únicos por columna:
names                  

### Distribución de variables numéricas

In [67]:
px.histogram(df_grades, x='math score', title='Distribución de "math score"', marginal='violin')


In [68]:
px.histogram(df_grades, x='reading score', title='Distribución de "reading score"', marginal='violin')

In [69]:
px.histogram(df_grades, x='writing score', title='Distribución de "writing score"', marginal='violin')

> Utilice alguna visualización (gráfique) para observar el comportamiento de los datos numericos. ¿Estan todos los valores correctos? 

No están todos los valores correctos, ya que una parte de ellos supera el valor máximo definido para los puntajes, es decir, tienen un valor mayor que 100.

**Justificación:**

```
Sí es necesario limpiar los datos ya que la base de datos presenta problemas de distinto tipo que pueden solucionarse mediante la limpieza. Estos problemas son la existencia de filas repetidas, valores nulos y valores fuera de los rangos permitidos para las variables numéricas. Esto último se observa claramente en las distribuciones mostradas, o también en la descripción, ya que el valor máximo supera el valor de 100 en los 3 casos.
```

### Análisis Univariado  

Observación de las distribuciones de las variables involucradas, tanto numéricas como categóricas.

In [90]:
numeric = ['math score', 'writing score', 'reading score']
categorical = [col for col in list(df_grades.columns) if col not in numeric]

In [122]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

#Análisis univariado de variables numéricas
fig = make_subplots(rows=1, cols=3, subplot_titles=numeric)

for idx, col in enumerate(numeric):
    row_idx = 1
    col_idx = idx+1

    data = df_grades.loc[:, col]
    hist = go.Histogram(x=data, name=col, histnorm="probability")
    fig.add_trace(hist, row=row_idx, col=col_idx)

fig.update_layout(
    height=400,
    title_text="Análisis Univariado de las Variables Numéricas",
    showlegend=False,
)
fig.show()


#Análisis univariado de variables categóricas
n_cols = 2
fig = make_subplots(rows=3, cols=n_cols, subplot_titles=categorical)

for idx, col in enumerate(categorical):
    row_idx = idx // n_cols + 1
    col_idx = idx % n_cols + 1

    data = df_grades.loc[:, col]
    hist = go.Histogram(x=data, name=col)
    fig.add_trace(hist, row=row_idx, col=col_idx)

fig.update_layout(
    height=1200,
    title_text="Análisis Univariado de las Variables Categoricas",
    showlegend=False,
)
fig.show()

### Análisis Bivariado  

Visualización de matrices de dispersión bivariadas de las variables numéricas, observando además la distribución de determinadas variables categóricas. En esta oportunidad, se realiza un análisis bivariado con respecto a la preparación previa, el género y el nivel educacional de los padres.

In [123]:
#Análisis bivariado de variables numéricas según preparación
fig = px.scatter_matrix(
    df_grades,
    dimensions=numeric,
    height=600,
    color='test preparation course',
    title="Análisis Bivariado de variables numéricas según preparación",
)
fig.update_traces(diagonal_visible=False)
fig.show()

#Análisis bivariado de variables numéricas según género
fig = px.scatter_matrix(
    df_grades,
    dimensions=numeric,
    height=600,
    color='gender',
    title="Análisis Bivariado de variables numéricas según género",
)
fig.update_traces(diagonal_visible=False)
fig.show()

#Análisis bivariado de variables numéricas según educación de los padres
fig = px.scatter_matrix(
    df_grades,
    dimensions=numeric,
    height=600,
    color='parental level of education',
    title="Análisis Bivariado de variables numéricas según educación de los padres",
)
fig.update_traces(diagonal_visible=False)
fig.show()

## 1.2 Limpieza de Datos [1 punto]

<p align="center">
  <img src="https://media3.giphy.com/media/c09yGbnSyaFRS/giphy.gif?cid=ecf05e47d0ejjjnpfj0ywcbwz7zk7sw2rfdpis890gmdpvj2&rid=giphy.gif&ct=g">
</p>


En base a lo realizado anteriormente, tome la decisión (si es necesario) de limpiar el dataset. Asuma los supuestos que estime conveniente para limpiar el DataFrame y cree una función que le permita limpiar los datos, si estima que no es necesario realizar esto, omita esta parte.

**To-Do:**

- [X] Decidir si realizar o no la limpieza.
- [X] Crear función que nos permita limpiar el dataset.
- [X] Justifique las asunciones realizadas para limpiar los datos.

**Hint**

En el caso que decida limpiar el dataset, puede que le sean de utilidad las funciones ```.drop_duplicates()``` y ```.dropna()```.


In [272]:
def limpieza_datos(df):
    #Eliminar filas repetidas
    df = df.drop_duplicates()
    
    #Eliminar nombres repetidos
    df = df.drop_duplicates(subset=['names'])
    
    #Enmascarar valores fuera de rango con NaNs
    df['math score'] = df['math score'].mask(df['math score'] > 100)
    df['reading score'] = df['reading score'].mask(df['reading score'] > 100)
    df['writing score'] = df['writing score'].mask(df['writing score'] > 100)
    
    #ELiminar filas en que 3 scores son NaN
    df = df.dropna(axis=0, subset=numeric, how='all') 
    
    return df

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

In [297]:
df_grades_clean = limpieza_datos(df_grades)
analisis_preliminar_datos(df_grades_clean)

--Fragmento del DataFrame:


names  gender race/ethnicity parental level of education  \
470        Evan Ammons    male        group D           bachelor's degree   
240     Judith Lindsay  female        group C           bachelor's degree   
688       Henry Rogers    male        group A            some high school   
901  Virginia Blubaugh  female        group B            some high school   
775       Arturo Brown    male        group A                 high school   

            lunch test preparation course  math score  reading score  \
470  free/reduced                    none        68.0           68.0   
240  free/reduced               completed        51.0           72.0   
688  free/reduced                    none        68.0           72.0   
901      standard               completed        60.0           70.0   
775      standard                    none        63.0           63.0   

     writing score  
470           67.0  
240           79.0  
688           64.0  
901           70.0  
775           62.0


--Dimensiones del DataFrame:
920 filas y 9 columnas

--Descripción de datos numéricos:
       math score  reading score  writing score
count  893.000000     892.000000     891.000000
mean    65.941769      69.103139      67.759820
std     15.306545      14.683053      15.415905
min      0.000000      17.000000      10.000000
25%     57.000000      59.000000      57.000000
50%     66.000000      70.000000      69.000000
75%     76.000000      79.000000      79.000000
max    100.000000     100.000000     100.000000

--Total de filas repetidas:
0

--Cantidad de valores nulos:
Por columna:
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     27
reading score                  28
writing score                  29
dtype: int64
Total: 84

--Cantidad de valores únicos por columna:
names                          920

Se verifica que los valores máximos no exceden los rangos permitidos. También se observa que la canntidad de valores nulos se redujo y que ya no hay filas repetidas. También podemos verificar lo primero revisando nuevamente la distribución de las variables numéricas.

In [285]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

#Análisis univariado de variables numéricas
fig = make_subplots(rows=1, cols=3, subplot_titles=numeric)

for idx, col in enumerate(numeric):
    row_idx = 1
    col_idx = idx+1

    data = df_grades_clean.loc[:, col]
    hist = go.Histogram(x=data, name=col, histnorm="probability")
    fig.add_trace(hist, row=row_idx, col=col_idx)

fig.update_layout(
    height=400,
    title_text="Análisis Univariado de las Variables Numéricas",
    showlegend=False,
)
fig.show()


**Justificación:**

```
Las asunciones realizadas son las siguientes: primero, no se consideró válida la existencia de filas repetidas, ya que esto corresponde a información redundante que no es un aporte a los datos. En segundo lugar, se asume que cada alumno rinde una única vez cada prueba, por lo que no pueden haber filas con el mismo atributo 'names', porque esto entregaría múltiples puntajes en las prueba para cada nombre. También se asumió como válido el hecho de que un alumno pueda haber rendido 1, 2 o 3 pruebas, por lo que se admite la existencia de NaNs en alguna de las columnas numéricas. No obstante, se consideró inválido el caso en que los 3 valores son NaN, ya que este caso no aporta información al dataset, ya sea porque el alumno no rindió las pruebas o por errores en los registros. Al tratarse de rendimientos individuales de los alumnos, se decidió no imputar los valores con el promedio del resto (o alguna otra técnica) y se optó por eliminar dichas filas.
```

## 1.3 Análisis exploratorio de datos



<p align="center">
  <img src="https://media4.giphy.com/media/ko7twHhomhk8E/giphy.gif?cid=ecf05e47xmgfz8ftaom8saa4cwktxpb0fkui9yk3nakb4ff1&rid=giphy.gif&ct=g">
</p>


### 1.3.1 Transformación de notas a "Nota Chilena" [0.5 punto]

En la ultima reunion con Don Caguayo este les comenta es complejo para el y su equipo entender la graduación de notas de 0 a 100. Por ende, le comenta que desde hace harto tiempo tienen la intención de *transformar* todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree un función que le permita pasar a la escala Chilena las notas que se escapan del rango [1,7]. 

(No ocupe ningún tipo de ciclo).

**To-Do**

- [X] Transforme los valores a la escala chilena.
- [X] Asigne los valores transformados al DataFrame.

In [298]:
def score_a_notachilena(score):
    nota = score*0.06 + 1
    return nota

In [300]:
df_grades_clean['math score'] = df_grades_clean['math score'].map(lambda score: score_a_notachilena(score))
df_grades_clean['reading score'] = df_grades_clean['reading score'].map(lambda score: score_a_notachilena(score))
df_grades_clean['writing score'] = df_grades_clean['writing score'].map(lambda score: score_a_notachilena(score))

In [302]:
analisis_preliminar_datos(df_grades_clean)

--Fragmento del DataFrame:


names  gender race/ethnicity parental level of education  \
842       Ofelia Evans  female        group C             master's degree   
794    Donald Mooberry    male        group D           bachelor's degree   
1030    Phillip Powell    male        group E                some college   
820      Nicole Obrien  female        group B           bachelor's degree   
350   Lawrence Cormier    male        group B          associate's degree   

             lunch test preparation course  math score  reading score  \
842       standard               completed        5.86           6.46   
794       standard               completed        5.08           5.44   
1030      standard               completed        5.86           5.44   
820   free/reduced                    none        5.62           6.10   
350       standard                    none        4.90           4.24   

      writing score  
842            6.22  
794            5.44  
1030           5.26  
820            6.22  
350            4.42


--Dimensiones del DataFrame:
920 filas y 9 columnas

--Descripción de datos numéricos:
       math score  reading score  writing score
count  893.000000     892.000000     891.000000
mean     4.956506       5.146188       5.065589
std      0.918393       0.880983       0.924954
min      1.000000       2.020000       1.600000
25%      4.420000       4.540000       4.420000
50%      4.960000       5.200000       5.140000
75%      5.560000       5.740000       5.740000
max      7.000000       7.000000       7.000000

--Total de filas repetidas:
0

--Cantidad de valores nulos:
Por columna:
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     27
reading score                  28
writing score                  29
dtype: int64
Total: 84

--Cantidad de valores únicos por columna:
names                          920

Se verifica que el valor máximo de las variables numéricas es 7.0.

### 1.3.2 Obtención de promedio de alumnos [0.5 puntos]

Con todas las notas adjuntadas, la prestigiosa *Universidad De la Cachaña* necesita obtener el promedio general de cada uno de los alumnos (Grade Point Average - GPA). Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del DataFrame original.

**To-Do:**
- [X] Obtenga el promedio de cada uno de los alumnos.
- [X] Anexe esta información en el DataFrame original.


In [307]:
#Calcular promedios
gpa_series = df_grades_clean[['math score','reading score','writing score']].mean(axis=1, skipna=True)

#Anexar a DF
df_grades_clean['GPA'] = gpa_series

In [320]:
df_grades_clean.sample(5)

names  gender race/ethnicity parental level of education  \
1352   Sandra Pompey  female        group E          associate's degree   
1362    Carol Miller  female        group C          associate's degree   
973   Shirley Orphey  female        group C                 high school   
432    Henry Mcgowen    male        group E                some college   
373      Lorene Hall  female        group D                some college   

             lunch test preparation course  math score  reading score  \
1352      standard                    none        7.00           7.00   
1362      standard               completed        5.50           5.92   
973   free/reduced                    none         NaN            NaN   
432       standard                    none        5.56           5.26   
373   free/reduced                    none        4.30           5.26   

      writing score   GPA  
1352           7.00  7.00  
1362           6.40  5.94  
973            4.84  4.84  
432            5.32  5.38  
373            5.14  4.90

### 1.3.3 Reprobados [1 punto]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena.

Cree una función que permita obtener dos Dataframes, el primero debe poseer solamente los alumnos que han reprobado en la prestigiosa universidad y el segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.

**To-Do:**

- [X] Obtenga una máscara DataFrame con los alumnos que reprobaron el año.
- [X] Reemplaze todos los GPA reprobados con una **R**.

In [325]:
#Máscara reprobados
mask_r = df_grades_clean['GPA'] < 4.0
df_reprobados = df_grades_clean[mask_r]
df_reprobados.sample(5)

names  gender race/ethnicity parental level of education  \
707  Harvey Pribbenow    male        group D           bachelor's degree   
89          Ann Pence  female        group B          associate's degree   
807    Arlene Seymour  female        group B                 high school   
954       Rosa Clagon  female        group A            some high school   
643     Lauren Warren  female        group C            some high school   

            lunch test preparation course  math score  reading score  \
707  free/reduced                    none        4.30           3.76   
89       standard                    none        3.40           3.88   
807      standard                    none        3.52           4.12   
954  free/reduced                    none        3.64           3.70   
643  free/reduced                    none        3.64           4.00   

     writing score   GPA  
707           3.64  3.90  
89            4.00  3.76  
807           4.06  3.90  
954           3.70  3.68  
643           4.06  3.90

In [327]:
#GPA reprobados por R
df_reprobados_gpa = df_grades_clean.copy()
df_reprobados_gpa.loc[mask_r, 'GPA'] = 'R'
df_reprobados_gpa.sample(5)

names  gender race/ethnicity parental level of education  \
124  Estelle Jackson  female        group E                 high school   
984   Jonathan Moser    male        group B                 high school   
994      Anne Scheer  female        group C                 high school   
201       Linda Lamp  female        group C            some high school   
148      John Echols    male        group D          associate's degree   

            lunch test preparation course  math score  reading score  \
124  free/reduced               completed        4.96           5.44   
984      standard                    none        4.12           3.88   
994      standard                    none        5.32           5.80   
201  free/reduced               completed        4.84            NaN   
148      standard                    none        6.40           6.22   

     writing score   GPA  
124           5.68  5.36  
984           3.94     R  
994           5.50  5.54  
201            NaN  4.84  
148           6.10  6.24

### 1.3.4 Información por Categorías [1 punto]

Motivado por saber más sobre los datos, el Rector les solicita que genere una función que le permita comparar los promedios de diferentes **atributos categóricos** del DataFrame. Para esto su función deberá tener como entrada el DataFrame con todos los promedios (sin agregar las R's) y calcular el promedio **agrupando por etnias y el nivel educacional de los padres**.

**To-Do:**

- [X] Hacer una función que nos permita obtener la comparación de promedios para dos atributos categóricos.
- [X] Cada uno de los grupos expuestos deben exponer el porcentaje de alumnos que representan en el colegio.
- [X] Comente lo que observa de los resultados.
- [X] Visualice los datos generados con [`px.sunburst`](https://plotly.com/python/sunburst-charts/ ).


In [390]:
def group_by_categorical(df):
    groupby = df.groupby(['race/ethnicity', 'parental level of education'])
    df_group = groupby.mean()
    df_group['porcentaje%'] = 100*groupby.size() / groupby.size().sum()
    return df_group

In [412]:
df_group = group_by_categorical(df_grades_clean)
display(df_group)

math score  reading score  \
race/ethnicity parental level of education                              
group A        associate's degree             4.660000       5.024286   
               bachelor's degree              4.840000       4.978000   
               high school                    4.631765       4.811765   
               master's degree                4.690000       5.230000   
               some college                   4.510000       4.752308   
               some high school               4.465714       4.618261   
group B        associate's degree             4.985714       5.165714   
               bachelor's degree              5.225263       5.446316   
               high school                    4.509268       4.744000   
               master's degree                4.912000       5.692000   
               some college                   4.808125       4.984706   
               some high school               4.731351       5.003333   
group C        associate's degree             4.990909       5.309091   
               bachelor's degree              5.027895       5.498421   
               high school                    4.616552       4.904211   
               master's degree                4.924000       5.144000   
               some college                   4.900000       5.156271   
               some high school               4.646667       4.930000   
group D        associate's degree             5.050612       5.238298   
               bachelor's degree              5.150435       5.144800   
               high school                    4.750698       4.841395   
               master's degree                5.288571       5.542000   
               some college                   5.126667       5.270476   
               some high school               5.057333       5.245000   
group E        associate's degree             5.512973       5.430000   
               bachelor's degree              5.683750       5.507500   
               high school                    5.456667       5.390000   
               master's degree                5.540000       6.030000   
               some college                   5.403571       5.315862   
               some high school               5.320000       5.173333   

                                            writing score       GPA  \
race/ethnicity parental level of education                            
group A        associate's degree                4.814286  4.832857   
               bachelor's degree                 4.996000  4.938000   
               high school                       4.681176  4.708235   
               master's degree                   5.350000  5.090000   
               some college                      4.664286  4.561333   
               some high school                  4.448800  4.544615   
group B        associate's degree                5.090286  5.080571   
               bachelor's degree                 5.380000  5.350526   
               high school                       4.600000  4.578095   
               master's degree                   5.548000  5.384000   
               some college                      4.767273  4.855429   
               some high school                  4.869143  4.849474   
group C        associate's degree                5.236923  5.158235   
               bachelor's degree                 5.507895  5.344737   
               high school                       4.755789  4.730508   
               master's degree                   5.104000  5.057333   
               some college                      5.074915  5.003226   
               some high school                  4.790909  4.765532   
group D        associate's degree                5.202553  5.135510   
               bachelor's degree                 5.257500  5.127200   
               high school                       4.766047  4.786047   
               master's degree                   5.734000  5.555238   
   

**Respuesta Esperada:**

| race/ethnicity | parental level of education | math score | reading score | writing score |  GPA | percentaje % |
|---------------:|----------------------------:|-----------:|--------------:|--------------:|-----:|-------------:|
|     group A    |      associate's degree     |       4.66 |          5.02 |          4.81 | 4.83 |          1.4 |
|                |      bachelor's degree      |       5.03 |          5.08 |          5.10 | 5.07 |          1.2 |
|                |         high school         |       4.63 |          4.77 |          4.63 | 4.68 |          1.8 |
|                |       master's degree       |       4.46 |          4.88 |          5.06 | 4.80 |          0.3 |
|                |         some college        |       4.83 |          4.95 |          4.90 | 4.89 |          1.8 |
|                |       some high school      |       4.54 |          4.72 |          4.51 | 4.59 |          2.4 |
|     group B    |      associate's degree     |       4.97 |          5.18 |          5.09 | 5.08 |          4.1 |
|                |      bachelor's degree      |       5.16 |          5.38 |          5.30 | 5.28 |          2.0 |
|                |         high school         |       4.59 |          4.81 |          4.68 | 4.69 |          4.8 |
|                |       master's degree       |       5.03 |          5.81 |          5.63 | 5.49 |          0.6 |
|                |         some college        |       4.79 |          4.95 |          4.85 | 4.86 |          3.7 |
|                |       some high school      |       4.71 |          4.99 |          4.88 | 4.86 |          3.8 |
|     group C    |      associate's degree     |       5.00 |          5.27 |          5.22 | 5.16 |          7.8 |
|                |      bachelor's degree      |       5.09 |          5.54 |          5.55 | 5.39 |          4.0 |
|                |         high school         |       4.65 |          4.87 |          4.70 | 4.74 |          6.4 |
|                |       master's degree       |       5.02 |          5.23 |          5.17 | 5.14 |          1.9 |
|                |         some college        |       4.91 |          5.17 |          5.13 | 5.07 |          6.9 |
|                |       some high school      |       4.63 |          4.94 |          4.80 | 4.79 |          4.9 |
|     group D    |      associate's degree     |       5.06 |          5.23 |          5.19 | 5.16 |          5.0 |
|                |      bachelor's degree      |       5.05 |          5.21 |          5.31 | 5.19 |          2.8 |
|                |         high school         |       4.77 |          4.86 |          4.79 | 4.81 |          4.4 |
|                |       master's degree       |       5.35 |          5.63 |          5.78 | 5.59 |          2.3 |
|                |         some college        |       5.12 |          5.25 |          5.30 | 5.23 |          6.7 |
|                |       some high school      |       5.01 |          5.20 |          5.15 | 5.12 |          5.0 |
|     group E    |      associate's degree     |       5.49 |          5.43 |          5.39 | 5.44 |          3.9 |
|                |      bachelor's degree      |       5.59 |          5.49 |          5.52 | 5.54 |          1.8 |
|                |         high school         |       5.25 |          5.22 |          5.05 | 5.17 |          2.2 |
|                |       master's degree       |       5.48 |          5.93 |          5.83 | 5.74 |          0.8 |
|                |         some college        |       5.43 |          5.36 |          5.21 | 5.33 |          3.5 |
|                |       some high school      |       5.33 |          5.17 |          4.99 | 5.16 |          1.8 |

> Visualice utilizando [`px.sunburst`](https://plotly.com/python/sunburst-charts/ ) el DataFrame obtenido en esta sección. 

In [397]:
px.sunburst(df_group.reset_index(), path=['race/ethnicity', 'parental level of education'], values='GPA')

> Comente los resultados obtenidos. ¿Hay alguna diferencia entre los grupos graficados?, ¿Es útil esta visualización?

**Justifique:**

```
Al observar los resultados obtenidos puede verse que al parecer existe una tendencia a que ciertos grupos étnico-raciales presenten mejores resultados que otros, para los distintos niveles eduacacionales de los padres. Al menos puede identificarse que el grupo E tiene un mejor GPA para todos los niveles, reflejando mejores resultados, mientras que el grupo A es en promedio el de peor desempeño. Además puede observarse que el nivel educacional de los padres también influye en el desempeño, notándose que la categoría "masters's degree" lidera en GPA para los 5 grupos.

Con respecto a la visualización, esta es útil para visualizar los resultados de forma interactiva y atractiva, pero no creo que signifique un gran aporte, ya que se pierden las nociones de comparación entre las distintas categorías y sus   GPA.
```

### 1.3.5 Combinar Dataset [1 punto]

Al notificarle los resultados obtenidos en la etapa de descubrimiento a Don Caguayo, este les informa que ha encontrado un Excel con nueva información, en esta se tienen los valores de dos asignaturas y desea que la unan con el dataset original (```students_grades.csv```).

Busque la forma de unir ambos DataFrames, para esto deberá anexar las columnas de ```history score``` y ```science score``` al final del DataFrame original.

**To-Do**

- [ ] Unir DataFrame ```students_grades.csv``` y ```other_grades.csv```.

In [415]:
df_other_grades = pd.read_csv(dir+'other_grades.csv')

In [418]:
pd.merge(left=df_grades, right=df_other_grades, on='names', how='inner')

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1        Rita Courtney  female        group B            some high school   
2     Charles Linstrom    male        group A           bachelor's degree   
3          Brian Young    male        group C            some high school   
4          Brian Young    male        group C            some high school   
...                ...     ...            ...                         ...   
1242     Richard Young    male        group D                 high school   
1243     Wanda Russell  female        group B                 high school   
1244    Marina Zeigler  female        group C           bachelor's degree   
1245     Laurie Carter  female        group B            some high school   
1246      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
1         standard                    none       193.0          193.0   
2         standard               completed        80.0           78.0   
3         standard                    none        73.0           66.0   
4         standard                    none         NaN            NaN   
...            ...                     ...         ...            ...   
1242      standard                    none        69.0           75.0   
1243  free/reduced               completed        23.0           44.0   
1244  free/reduced               completed        66.0           74.0   
1245      standard               completed        54.0           61.0   
1246      standard               completed        68.0           80.0   

      writing score  science score  history score  
0              46.0             25            3.8  
1             193.0             25            3.8  
2              81.0             56            5.2  
3              63.0             58            3.8  
4               NaN             58            3.8  
...             ...            ...            ...  
1242           71.0             62            4.2  
1243           36.0             16            3.6  
1244           81.0             59            4.1  
1245           62.0             43            4.3  
1246           76.0             61            5.3  

[1247 rows x 11 columns]

> Explique (pensando en una operación de conjuntos) que hace la operación de combinación por defecto (`inner`).

> Explique que hubiese sucedido si aplicaramos los siguientes merge en las variables anteriores:

- `left`
- `right`
- `outer`

**Justificación:**

```
La operación por defecto "inner" combina los elementos de ambos dataset conservando la intersección (pensándolo como conjuntos). Es decir, conserva los elementos que se encuentran en ambas tablas, descartando los que no encuentran su par. 

Si se aplica "left", además de la intersección se conservan las filas del primer dataset que no se combinaron con ninguna fila del segundo dataset.

Al aplicar "right" ocurre lo análogo a lo anterior, pero conservando las filas del segundo dataset.

Con "outer" ocurre que además de la intesección, se almacenan todas las filas de ambos dataset que no fueron combinadas.
```

## 1.4 Más visualizaciones [Bonus] [1 punto]


**Desde aquí parte el bonus para el lab.**


<p align="center">
  <img src="https://media4.giphy.com/media/xT5LMWNOjGqJzUfyve/giphy.gif?cid=ecf05e47phkhvg6vadqvcoobvxwj9bzxxk38jjvkcc6eak6t&rid=giphy.gif&ct=g">
</p>




### 1.4.1 Rendimiento por Asignatura

Con el DataFrame de rendimiento para cada uno de los alumnos grafique como distribuyen los datos para cada uno de los puntajes obtenidos en cada una de las asignaturas.

Para realizar esta tarea se les recomienda que utilicen el comando ```px.histogram``` de plotly. Para esto puede ser útil procesar los datos con un melt, agrupando en una columna las asignaturas y en otra columna la nota obtenida en dicha asignatura.

Ejemplo del melt:

|    Variable   | Value |
|:-------------:|:-----:|
| math score    | 6.4   |
| math score    | 5.2   |
| ...           | ...   |
| writing score | 4.5   |

**To-Do:**

- [X] Utilice `plotly` para visualizar el rendimiento para los diferentes grupos que componen el dataset.
- [X] Responder a la pregunta, ¿Cuál es la asignatura que más les cuesta a los alumnos de la universidad de la Cachaña?


In [440]:
melt = df_grades_clean[numeric].melt()
px.histogram(melt, color='variable')

> ¿Cuál es la asignatura que más les cuesta a los alumnos de la universidad de la Cachaña?

**Comente:**

```
Según lo observado en las distribuciones, la asignatura que más le cuesta a los alumnos es Matemáticas, ya que los valores de "math score" parecen mostrar un valor promedio levemente más bajo que el del resto de las asignaturas, además de una cola más larga hacia las bajas calificaciones.
```

### 1.4.2 Otras Visualizaciones

Genere dos visualizaciones extras que encuentre interesantes con estos datos. Agrupe los atributos que estime convenientes.


**To-Do:**
- [X] Generar dos nuevas visualizaciones con los datos.

In [450]:
#Análisis bivariado de variables numéricas según educación de los padres
fig = px.scatter_matrix(
    df_grades_clean,
    dimensions=numeric,
    height=600,
    color='lunch',
    title="Análisis Bivariado de variables numéricas según tipo de almuerzo",
)
fig.update_traces(diagonal_visible=False)
fig.show()

Del análisis bivariado de variables numéricas según tipo de almuerzo es llamativo el hecho de que, en todos los casos, los peores puntajes corresponden a personas con un almuerzo de tipo "free/reduced".

In [451]:
corrmat = df_grades_clean[numeric].corr()
px.imshow(corrmat, title='Matriz de correlación de puntajes')

Al observar la matriz de correlación, es notorio que el rendimiento en las pruebas de lectura y escritura tienen una mayor correlación entre sí que la que hay entre ellas y los puntajes de matemáticas, lo cual tiene sentido.

# Conclusión
Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>

<br>
<center>
<img src="https://i.kym-cdn.com/photos/images/original/001/194/195/b18.png" width=100 height=50 />
</center>
<br>